### PR TITLE
Token bucket PFCP Session Report limiter

### PR DESCRIPTION
**Pull request by sergeymatov**
Generic rate limiter based on vnet/policer

### CODE REVIEW

File: upf/upf.c [hunk 1]
```diff
@@ -195,6 +195,24 @@ vnet_upf_nat_pool_add_del (u8 * nwi_name, ip4_address_t start_addr,
   return 0;
 }
 
+void
+upf_pfcp_policer_init (upf_main_t * vm)
```
Comment (ivan4th) [inline upf/upf.c:199]: `upf_main_t *` vars are (historically) called `gtm` or sometimes `sm`, but never `vm` which is used for `vlib_main_t *`. Perhaps someday we should rename them to `um`, but at least let's not make the naming even more diverse

File: upf/upf.c [hunk 2]
```diff
@@ -195,6 +195,24 @@ vnet_upf_nat_pool_add_del (u8 * nwi_name, ip4_address_t start_addr,
...
+
+  policer_main.single_rate = 1;
+  /* Tokens per period, according to police.h should be > 1000 */
+  policer_main.cir_tokens_per_period = 1 << 10;
+  /* scale is set to 10 for init, with given logic that message size is 1 */
+  policer_main.scale = 10;
+  /* we are making top limit of bucket to be high by default */
+  policer_main.current_limit = (1 << 31);
+
+  vec_add1 (vm->upf_pfcp_policer_main, policer_main);
```
Comment (ivan4th) [inline upf/upf.c:213]: why use a vector here? that's just a single instance. Perhaps could just have a global var (which is already defined below in this commit btw)

File: upf/upf_cli.c [hunk 3]
```diff
@@ -174,6 +174,38 @@ VLIB_CLI_COMMAND (upf_pfcp_show_endpoint_command, static) =
 };
 /* *INDENT-ON* */
 
+static clib_error_t *
+upf_pfcp_policer_set_fn (vlib_main_t * vm,
+			 unformat_input_t * main_input,
+			 vlib_cli_command_t * cmd)
```
Comment (ivan4th) [inline upf/upf_cli.c:180]: I don't see a definition of CLI command that uses this func. Also, binapi would be nice

**Review by ivan4th — CHANGES_REQUESTED** [review]

File: upf/upf.c [hunk 4]
```diff
@@ -195,6 +195,23 @@ vnet_upf_nat_pool_add_del (u8 * nwi_name, ip4_address_t start_addr,
...
+
+  policer_t *policer_main = &upf_pfcp_policer_main;
+
+  policer_main->single_rate = 1;
+  /* Tokens per period, according to police.h should be > 1000 */
+  policer_main->cir_tokens_per_period = 1 << 10;
+  /* scale is set to 10 for init, with given logic that message size is 1 */
+  policer_main->scale = 10;
+  /* we are making top limit of bucket to be high by default */
+  policer_main->current_limit = (1 << 31);
```
Comment (RoadRunnr) [inline upf/upf.c:210]: I'm not sure that we should init the policer like this. The normal API seems to use a qos_pol_cfg_params_st struct and pol_logical_2_physical to translate.

File: upf/upf_pfcp_server.c [hunk 5]
```diff
@@ -407,6 +407,13 @@ upf_pfcp_server_send_session_request (upf_session_t * sx,
 				      pfcp_decoded_msg_t * dmsg)
 {
   pfcp_msg_t *msg;
+  upf_main_t *gtm = &upf_main;
+  policer_t *policer_main = gtm->upf_pfcp_policer_main;
+
+  if (vnet_police_packet
+      (policer_main, 1, POLICE_VIOLATE,
```
Comment (RoadRunnr) [inline upf/upf_pfcp_server.c:414]: is think that should `POLICE_CONFORM`

**Review by RoadRunnr — CHANGES_REQUESTED** [review]

File: upf/upf.c (hunk 1 above)
Comment (sergeymatov) [reply upf/upf.c]: this thing has been refactored. Probably another review of the PR is required.

File: upf/upf.c (hunk 2 above)
Comment (sergeymatov) [reply upf/upf.c]: I've refactored PR to use singleton global QoS config that can be configurable via API or CLI
Configuration done via committed PPS and committed burst in ms.
@ivan4th @RoadRunnr please recheck the PR

**Review by ivan4th — APPROVED** [review]
It would be nice to have at least some test cases for that, even if just for the binapi itself (although verifying the policier with start-of-traffic reporting trigger is perhaps possible, too).
Other than that, LGTM

**Review by RoadRunnr — APPROVED** [review]

**Pull request merged**